### DISCUSSION TITLE
ModelMultipleChoiceFilter's custom method gets called even when field is not filtered on

### DISCUSSION THREAD

**Discussion by johncmacy** [carltongibson/django-filter#1622] [Q&A]
Hi,

I have a FilterSet with a ModelMultipleChoiceFilter that has a custom method:

``` py
class Filters(filters.FilterSet):
    publisher = filters.ModelMultipleChoiceFilter(
        queryset=Publisher.objects.all(),
        method="publisher_filter",
    )

    def publisher_filter(self, queryset, name, value):
        return queryset.filter(publisher__in=value)        # contrived example, we have more complex business logic
```

After getting some unexpected results, I found out that the custom method is getting called even when the field is not passed as a query param. For example, requests to `/api/book` return an empty list `[]`.

After digging into it a bit, it seems that _every_ filter gets called regardless if it's passed as a query param. Normally this is fine because they get an empty value and it doesn't apply any filtering. However, when a custom method is supplied, it gets called with `<QuerySet []>` as the `value`. So in the example above, it filters books by publisher in `[]`, which results in the empty result set.

I got around it with:

``` py
    def publisher_filter(self, queryset, name, value):
        if not value:
            return queryset
        return queryset.filter(publisher__in=value)
```

However, I didn't expect the method to get called in the first place, since the request didn't have a "publisher" query param. Am I missing something in the filter configuration that would prevent it from being called in the first place?

Thanks!

**Comment by carltongibson** [comment]
> ... every filter gets called regardless if it's passed as a query param. 

Filters are called if the field is present in `filterset.form.cleaned_data`. I'd need to dig in to see exactly why non-submitted fields are showing there. 🤔

It's **not** expected though... 

> Note that the value is validated by the Filter.field, so raw value transformation and empty value checking should be unnecessary.
> -- https://django-filter.readthedocs.io/en/stable/ref/filters.html#method

**Reply by carltongibson** [reply]
Created #1625 to investigate.
